### PR TITLE
SNOW-1672573 Add coverage for `DataFrame.to_local_iterator`

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
Fixes SNOW-1672573

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

- I added some error messages in the general AST testing code to make it easier to figure out if the required JAR file is missing. 
- I also modified a part of the mock connector code because it was returning without the required AST information.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_connection.py [hunk 1]
```diff
@@ -671,7 +668,7 @@ def execute(
 
         self.notify_query_listeners(QueryRecord("MOCK", "MOCK-PLAN"), **notify_kwargs)
 
-        return rows
+        return iter(rows) if to_iter else rows
```
Comment (sfc-gh-vbudati) [inline src/snowflake/snowpark/mock/_connection.py:671]: mock connection change -- previously it was returning before any of the AST data was sent.

File: tests/ast/conftest.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ def pytest_configure(config):
     pytest.unparser_jar = config.getoption("--unparser-jar")
     if not os.path.exists(pytest.unparser_jar):
         pytest.unparser_jar = None
+        logging.error(f"Unparser JAR not found at {pytest.unparser_jar}")
```
Comment (sfc-gh-vbudati) [inline tests/ast/conftest.py:38]: Error when trying to run `--update-expectations` but no jar file is found. When no jar file is found, the expected unparser output is updated to an empty string in every file.

File: tests/ast/data/DataFrame.to_local_iterator.test [hunk 3]
```diff
@@ -0,0 +1,32 @@
+## TEST CASE
+
+df = session.table("test_table")
```
Comment (sfc-gh-vbudati) [inline tests/ast/data/DataFrame.to_local_iterator.test:3]: This is the relevant test for the PR, the rest of the tests are updating the base64 string generated.

File: tests/ast/test_ast_driver.py [hunk 4]
```diff
@@ -254,6 +254,8 @@ def run_test(session):
             r"SNOWPARK_TEMP_TABLE_(\w+)", "SNOWPARK_TEMP_TABLE_xxx", raw_unparser_output
         )
         return unparser_output, "\n".join(base64_batches)
+    except Exception as e:
```
Comment (sfc-gh-vbudati) [inline tests/ast/test_ast_driver.py:257]: Explicitly check for exceptions in the `try-finally` block since if an error is raised, it is completely ignored by the code.

File: src/snowflake/snowpark/dataframe.py [hunk 5]
```diff
@@ -834,8 +834,28 @@ def to_local_iterator(
...
+                    t = expr.statement_params.add()
+                    t._1 = k
+                    t._2 = v
+            expr.block = block
+            expr.case_sensitive = case_sensitive
+
+            self._session._ast_batch.eval(repr)
+
+            # Flush the AST and encode it as part of the query.
+            _, kwargs["_dataframe_ast"] = self._session._ast_batch.flush()
```
Comment (sfc-gh-evandenberg) [inline src/snowflake/snowpark/dataframe.py:858]: How is the result of the eval to be processed?  I'm assuming that is out of scope for this pr since we don't have end to end execution of the AST in place to return a result.  Is the idea we would write the results to a temp table and a local iterator implementation to stream the results?

File: tests/ast/data/DataFrame.to_local_iterator.test [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+rows = [row for row in df.to_local_iterator(statement_params={"SF_PARTNER": "FAKE_PARTNER"})]
+
+rows = [row for row in df.to_local_iterator(statement_params={"SF_PARTNER": "FAKE_PARTNER"}, case_sensitive=False)]
+
+
+## EXPECTED UNPARSER OUTPUT
+
+df = session.table("test_table")
+
+df.to_local_iterator()
```
Comment (sfc-gh-evandenberg) [inline tests/ast/data/DataFrame.to_local_iterator.test:18]: Do we only generate a result assignment if it's a dataframe output of the api call?

File: tests/ast/data/DataFrame.agg.test [hunk 7]
```diff
@@ -22,4 +22,4 @@ df4 = df.agg({"a": "count", "b": "sum"})
...
...c3RkZGV2X3BvcBIgCDsQXhoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAnKF4afXp7Ci0KKwoHCgUKA2NvbBIgCDoQXhoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAyKF4aKPoLJQogCDoQXhoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAyKF4SAWEiIAg6EF4aFlNSQ19QT1NJVElPTl9URVNUX01PREUgMiheIiAIOxBeGhZTUkNfUE9TSVRJT05fVEVTVF9NT0RFICcoXhABGiAIPBBeGhZTUkNfUE9TSVRJT05fVEVTVF9NT0RFIA4oXhIFCgNkZjIYAiICCAIKwQIKvgIKrgLaBaoCCgeKAgQKAggBEvwBCnuiDHgKIAgwEGAaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihgEij6CyUKIAgwEGAaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihgEgFhEir6CycKIAgwEGAaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihgEgNtaW4Ke6IMeAogCDAQYBoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAOKGASKPoLJQogCDAQYBoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAOKGASAWISKvoLJwogCDAQYBoWU1JDX1BPU0lUSU9OX1RFU1RfTU9ERSAOKGASA21heBABGiAIMBBgGhZTUkNfUE9TSVRJT05fVEVTVF9NT0RFIA4oYBIFCgNkZjMYAyICCAMKogIKnwIKjwLaBYsCCgeKAgQKAggBEt0BCtgB4gPUAQpYEij6CyUKIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEgFhEiz6CykKIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEgVjb3VudApWEij6CyUKIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEgFiEir6CycKIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEgNzdW0SIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEAEaIAgwEGIaFlNSQ19QT1NJVElPTl9URVNUX01PREUgDihiEgUKA2RmNBgEIgIIBBABGhESDwoNCgVmaW5hbBADGAsgCSIEEAEYFQ==
```
[the remark this answers is not in the document]
Comment (sfc-gh-vbudati) [reply tests/ast/data/DataFrame.agg.test:25]: Protobuf serialization is non-deterministic (cf. https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510)
These are the results I had when serializing protobuf. I ran the tests with the `--update-expectations` flag which is why it changed all of the tests.

File: src/snowflake/snowpark/dataframe.py (hunk 5 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/dataframe.py:858]: I'm not sure what the solution we decided on for this is. Writing the results to a temp table and having a local iterator implementation sounds like the right solution.

File: tests/ast/data/DataFrame.to_local_iterator.test (hunk 6 above)
Comment (sfc-gh-vbudati) [reply tests/ast/data/DataFrame.to_local_iterator.test:18]: I'm not sure what you mean by this

File: src/snowflake/snowpark/mock/_connection.py (hunk 1 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/mock/_connection.py:671]: this change is done for both a.) mock and b.) regular connection>

File: tests/ast/conftest.py (hunk 2 above)
Comment (sfc-gh-lspiegelberg) [reply tests/ast/conftest.py:38]: nit: Add `.` at end of sentence. Maybe add explanation/action user should take to fix this error.

File: tests/ast/data/DataFrame.agg.test (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfc-gh-lspiegelberg) [reply tests/ast/data/DataFrame.agg.test:25]: unfortunately yes, we expect these files to change here quite frequently. However, the PR should pass the expectation test merge gate.

File: tests/ast/data/DataFrame.to_local_iterator.test [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+rows = [row for row in df.to_local_iterator(statement_params={"SF_PARTNER": "FAKE_PARTNER"})]
+
+rows = [row for row in df.to_local_iterator(statement_params={"SF_PARTNER": "FAKE_PARTNER"}, case_sensitive=False)]
+
+
+## EXPECTED UNPARSER OUTPUT
+
+df = session.table("test_table")
+
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/data/DataFrame.to_local_iterator.test:17]: can we add one case with some other statement in between? I.e., a select or filter? Just to be sure edge cases are caught.

File: tests/ast/test_ast_driver.py (hunk 4 above)
Comment (sfc-gh-lspiegelberg) [reply tests/ast/test_ast_driver.py:257]: mabe let's add a comment, or reraise with some meaningful exception?

E.g.,
```
raise Exception('Generated AST test failed') from e
```
or so

File: tests/ast/test_ast_driver.py [hunk 9]
```diff
@@ -277,6 +279,9 @@ def test_ast(session, test_case):
         session, test_case.filename.replace(".", "_"), test_case.source
     )
     if pytest.update_expectations:
+        assert (
+            pytest.unparser_jar
+        ), "Can only update expectations with unparser jar set."
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/test_ast_driver.py:284]: nit: add actionable inside for user, i.e. some hint on what to do in order to fix it.

**Review by sfc-gh-lspiegelberg — APPROVED** [review]
lgtm, one thing to look out for is whether the fix for mock connection has also been applied to the real connection. You can easily check by running some integ test involving to_local_iterator, e.g. `test_dataframe_suite.py::test_to_local_iterator`.

File: tests/ast/test_ast_driver.py (hunk 9 above)
Comment (sfc-gh-vbudati) [reply tests/ast/test_ast_driver.py]: done! is the new message good or should I include more details?

File: tests/ast/data/DataFrame.to_local_iterator.test (hunk 8 above)
Comment (sfc-gh-vbudati) [reply tests/ast/data/DataFrame.to_local_iterator.test:21]: done!

File: src/snowflake/snowpark/dataframe.py [hunk 10]
```diff
@@ -834,8 +834,28 @@ def to_local_iterator(
             case_sensitive: A bool value which controls the case sensitivity of the fields in the
                 :class:`Row` objects returned by the ``to_local_iterator``. Defaults to ``True``.
         """
+
+        kwargs = {}
         if _emit_ast:
-            raise NotImplementedError("TODO SNOW-1672573: Support to_local_iterator.")
+            # Add an Assign node that applies SpDataframeToLocalIterator() to the input, followed by its Eval.
+            repr = self._session._ast_batch.assign()
```
[the remark this answers is not in the document]
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/dataframe.py]: switched it!

**Review by sfc-gh-evandenberg — APPROVED** [review]
LGTM
